core/dbt/parser/unit_tests.py

```diff
@@ -389,6 +391,44 @@ def _validate_and_normalize_rows(self, ut_fixture, unit_test_definition, fixture
                     ut_fixture.fixture, self.project.project_name, unit_test_definition.unique_id
                 )
 
+        # sanitize order of input
+        if ut_fixture.rows and (
+            ut_fixture.format == UnitTestFormat.Dict or ut_fixture.format == UnitTestFormat.CSV
+        ):
+            self._promote_first_non_none_row(ut_fixture)
+
+    def _promote_first_non_none_row(self, ut_fixture):
+        """
+        Promote the first row with no None values to the top of the ut_fixture.rows list.
+
+        This function modifies the ut_fixture object in place.
+
+        Needed for databases like Redshift which uses the first value in a column to determine
+        the column type. If the first value is None, the type is assumed to be VARCHAR(1).
+        This leads to obscure type mismatch errors centered on a unit test fixture's `expect`.
+        See https://github.com/dbt-labs/dbt-redshift/issues/821 for more info.
+        """
+        non_none_row_index = None
+
+        # Iterate through each row and its index
+        for index, row in enumerate(ut_fixture.rows):
+            # Check if all values in the row are not None
+            if all(value is not None for value in row.values()):
+                non_none_row_index = index
+                break
+
+        if non_none_row_index is None:
+            fire_event(
+                SystemStdErr(
+                    bmsg="Unit Test fixtures benefit from having at least one row free of Null values to ensure consistent column types. Failure to meet this recommendation can result in type mismatch errors between unit test source models and `expected` fixtures."
```

Comment:
this is very clear from my perspective reading this diff but I'm worried it's somewhat verbose for an end-user reading this. What about something more straightforward like: 

"Unit test fixutre <fixture-name> in <unit-test-name> does not have any row free of null values, which may cause type mismatch errors during unit test execution."

Reply:
I'll take this suggestion for the user facing exception to be raised in redshift